lib/constructs/callback-function.js

```diff
@@ -0,0 +1,102 @@
+"use strict";
+
+const Types = require("../types");
+const utils = require("../utils");
+
+class CallbackFunction {
+  constructor(ctx, idl) {
+    this.ctx = ctx;
+    this.idl = idl;
+    this.name = idl.name;
+    this.str = null;
+
+    this.requires = new utils.RequiresMap(ctx);
+  }
+
+  generate() {
+    // This is a simplification of https://heycam.github.io/webidl/#web-idl-arguments-list-converting that currently
+    // fits our needs.
+    const argsToES = `
+      for (let i = 0; i < args.length; i++) {
+        args[i] = utils.tryWrapperForImpl(args[i]);
+      }
+      if (args.length < ${this.idl.arguments.length}) {
+        for (let i = args.length; i < ${this.idl.arguments.length}; i++) {
+          args[i] = undefined;
+        }
+      } else if (args.length > ${this.idl.arguments.length}) {
+        args.length = ${this.idl.arguments.length};
+      }
+    `;
+
+    let returnIDL;
+
+    if (this.idl.idlType.idlType === "void") {
+      returnIDL = "";
+    } else {
+      const conv = Types.generateTypeConversion(this.ctx, "returned", this.idl.idlType, [], this.name, "context");
+      this.requires.merge(conv.requires);
+      returnIDL = `
+        ${conv.body}
+        return returned;
+      `;
+    }
+
+    const treatNonObjectAsNull = utils.getExtAttr(this.idl.extAttrs, "TreatNonObjectAsNull");
+    const assertCallable = treatNonObjectAsNull ? "" : `
+      if (typeof value !== "function" && !treatNonObjectAsNull) {
+        throw new TypeError(context + " is not a function");
+      }
+    `;
+
+    this.str += `
+      module.exports = {
+        convert(value, { context = "The provided value" } = {}) {
+          ${assertCallable}
+
+          function wrapper(...args) {
+            if (new.target !== undefined) {
+              throw new Error("Internal error: do not call converted callback function with \`new\`; instead use " +
+                              "convertedFunc.construct()");
+            }
+
+            let returned;
+    `;
+    if (treatNonObjectAsNull) {
+      this.str += `if (typeof value === "function") {`;
+    }
+    this.str += `
+            ${argsToES}
+            returned = Reflect.apply(value, this, args);
+    `;
+    if (treatNonObjectAsNull) {
+      this.str += "}";
+    }
+    this.str += `
+            ${returnIDL}
+          }
+          wrapper.construct = function(...args) {
+            ${argsToES}
+            let returned = Reflect.construct(value, args);
+            ${returnIDL}
```

Comment:
This&nbsp;should&nbsp;probably also&nbsp;be&nbsp;wrapped in&nbsp;<code>if&nbsp;(typeof&nbsp;value&nbsp;===&nbsp;"function")</code> when&nbsp;<code>treatNonObjectAsNull&nbsp;===&nbsp;true</code>.

Reply:
Actually, I just had a read through [the relevant spec](https://heycam.github.io/webidl/#construct-a-callback-function), and `[TreatNonObjectAsNull]` isn’t specified for constructor callback functions.